ntoskrnl/config/ntapi.c

```diff
@@ -1937,20 +1939,37 @@ NtUnloadKey2(IN POBJECT_ATTRIBUTES TargetKey,
     }
 
     /* Call the internal API. Note that CmUnloadKey() unlocks the registry only on success. */
-    Status = CmUnloadKey(KeyBody->KeyControlBlock, Flags);
+    Status = CmUnloadKey(KeyBody->KeyControlBlock, Flags, LockExHeld);
 
     /* Check if we failed, but really need to succeed */
-    if ((Status == STATUS_CANNOT_DELETE) && (Flags == REG_FORCE_UNLOAD))
-    {
-        /* TODO: We should perform another attempt here */
-        _SEH2_TRY
-        {
-            DPRINT1("NtUnloadKey2(%wZ): We want to force-unload the hive but couldn't unload it: Retrying is UNIMPLEMENTED!\n", TargetKey->ObjectName);
-        }
-        _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+    if ((Status == STATUS_CANNOT_DELETE) && (Flags != REG_FORCE_UNLOAD))
+    {
+        /*
+         * We failed to unload the hive normally, this usually happens
+         * when we could not safely remove empty cache entries.
+         */
+        CmpReleaseTwoKcbLockByKey(ChildConv, ParentConv);
+        ExReleasePushLockExclusive(&CmpLoadHiveLock);
+        CmpUnlockRegistry();
+
+        /* Lock the entire registry down, at the cost of performance */
+        CmpLockRegistryExclusive();
+        ExAcquirePushLockExclusive(&CmpLoadHiveLock);
+        LockExHeld = TRUE;
+
+        /*
+         * The KCB may have been deleted behind us while we were
+         * toying around with the registry locks. Consider
+         * ourselves as job done.
```

Comment:
The last sentence doesn't make sense in english. It should most likely read: `Consider our job done.`